### ISSUE TITLE
[tracking-ish issue] Doc pages are very slow to render

### ISSUE THREAD

**Issue by GuillaumeGomez** [rust-lang/rust#55900]
Message from [discord](https://discordapp.com/channels/442252698964721669/443492145567891458):

> the https://doc.rust-lang.org/std/iter/trait.Iterator.html page(and only that page?) seems to have a broken search bar that manifests as infinite? loading and high CPU usage in both Chrome and firefox, across stable, beta, nightly, and local versions.

Related:

- [x] [Minification makes JS parsing slower](https://github.com/rust-lang/rust/issues/57754)
- [ ] [Open
search-index.js slows down documentation browsing](https://github.com/rust-lang/rust/issues/56545)
- [ ] Others?

**Comment by ollie27** [comment]
The HTML for that page is currently about 12.5MB thanks to #51885. We're now rendering a copy of every trait method for every impl. We could revert #51885 or only render methods that have documentation in the impl.

**Comment by JeanMertz** [comment]
Just to confirm this issue: I wanted to add that even loading that page is next to impossible on Safari. It takes a very long time, and after it loads, scrolling is so slow that the page is next to useless. I often have to switch to Firefox (my secondary browser) only to read this specific page in the docs.

**Comment by crumblingstatue** [comment]
Not only is it very slow to render initially, but due to the sheer amount of content on the page, it is also very sluggish to scroll through it on my system.

I also agree that we should show less information by default. A 12.5 MB doc page for a single item sounds ridiculous.

I'd very much prefer if implementors were just listed plainly like so: `Implementors: std::ascii::EscapeDefault, EscapeDebug, std::char::EscapeDefault, EscapeUnicode, ToLowercase, ToUppercase, ...`, with them being links so you can click on them to go to their doc page.

This amount of duplication is a lot of wasted data. It makes the documentation bloated and slow.

**Comment by Stargateur** [comment]
Also it would be good to fix this as fast as possible, Iterator are a lot used and be able to read the doc is primordial.

Freeze on Chrome Version 70.0.3538.110 (Build officiel) (64 bits)
Load quite fast on Firefox 63.0.3 (64 bits) ;)

**Comment by GuillaumeGomez** [comment]
The PR is open and waiting for reviews.

**Comment by nocduro** [comment]
Has this been deployed? I'm still having really sluggish performance on https://doc.rust-lang.org/std/iter/trait.Iterator.html 

Related issue #57281

**Comment by steveklabnik** [comment]
I don't think it's made its way to stable, check the beta/nightly docs.

**Comment by nocduro** [comment]
Just tried beta and nightly in incognito and it works about the same. The page doesn't freeze or anything, but scrolling is a bit laggy. Chrome uses ~30-40% cpu when scrolling.

I think this patch is live though, because nightly shows some loading text while the sections are loaded instead of instantly trying to render?

**Comment by Stargateur** [comment]
yeah, that doesn't freeze but the page still load tons of information, is it possible to have the page that only load what user required, for example, a plus bottom that will only load content when opened ? That would allow to see method, without load everything.

Also, all theses informations are them really needed ? Why did we do this change ?

**Comment by GuillaumeGomez** [comment]
The rendering was slow in any case so whatever changes we want to perform afterwards, we'll still gain from them. We're debating about what information could/should be removed on the pages alongside making the HTML/JS lighter. It's moving forward.

**Comment by crumblingstatue** [comment]
I just checked the nightly doc page. On my computer, it's just as slow as it was before. It takes quite a lot of seconds before it loads, and even then the scrolling is sluggish.

**Comment by GuillaumeGomez** [comment]
I don't see such behavior (either on firefox or chrome) so the problem could come from your computer configuration?

**Comment by crumblingstatue** [comment]
> I don't see such behavior (either on firefox or chrome) so the problem could come from your computer configuration?

I have a pretty old computer, but reading documentation shouldn't require a beefy computer. No other documentation I ever used had this problem. In fact, old versions of the Iterator page posed no problem either. But the new versions are way too large.

**Comment by JeanMertz** [comment]
I know this is being worked on, so my post is just "another datapoint". Ignore this if it doesn't add anything you don't already know.

Are we talking about this page? https://doc.rust-lang.org/nightly/std/iter/trait.Iterator.html

If so, that one is (still) slow for me as well, on a 2017 MBP. It takes about 5 seconds to load, and once it does, scrolling down is jerky from time to time (but not as often as before), and shows white screens until rendering catches up.

I do believe it's quite a bit faster than before, but still not great.

It's also worth considering what information people want to get from that page. Anything above "Implementations on Foreign Types" is only a fraction of the contents of the page (understandably, as the rest is mostly about other implementors), while it contains all you need to know on _how_ to use `Iterator` (what that section lacks, is _what_ implements Iterator, which is the bulk of the page) so perhaps it makes sense to put a cut there, and load the rest of the page lazily on request.

**Comment by Stargateur** [comment]
there is no need to prove the page is slow:

```
curl "https://doc.rust-lang.org/nightly/std/iter/trait.Iterator.html" > /dev/null
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 13.1M  100 13.1M    0     0  3375k      0  0:00:04  0:00:04 --:--:-- 2761k
```

13.1M, for a html page ? there is a problem. Even, C11 standard only does 2M, https://port70.net/~nsz/c/c11/n1570.html, and it already too big.

**Comment by GuillaumeGomez** [comment]
We agree on this and minification is being worked on like said previously: https://github.com/rust-lang/rust/pull/56869

**Comment by Stargateur** [comment]
I don't think micro optimization will fix the problem, we need a real solution that change how doc is working. 13 - 1 => 12 M. That will not change much. Also, that not a problem for me to download heavy thing, the problem is to render too many thing in the page.

We could argue that the problem is "only" in iterator, but iterator just show there is a problem, that could happen.

**Comment by DianaNites** [comment]
Wouldn't a better fix would be for pages to just not to repeat their own trait documentation, just a list of what implements it and for what `Item`?(Or more generally whatever associated constants) Cut it out entirely.

Then the Iterator page wouldn't repeat it's entire documentation for everything. That seems to me to be where the bulk of the page is.

**Comment by 0xpr03** [comment]
The iter site works pretty fine on my side.
Yeah it takes some seconds but scrolling is fine on my laptop. I don't think we can do too much about this without removing relevant information. Sometimes it's just that: A pretty huge amount of stuff to display. Just as in high LoC files.

**Comment by GuillaumeGomez** [comment]
> I don't think micro optimization will fix the problem, we need a real solution that change how doc is working. 13 - 1 => 12 M.

This is just one amongst others. There are PRs to reduce DOM size and other things. When we'll reach the end of those optimizations, if it's still not enough, we'll check for other solutions at this time.

**Comment by steveklabnik** [comment]
There's always tension between different use-cases. Sometimes, people want docs for individual implementations, to talk about what they do, etc. That can only happen if things are repeated.

**Comment by DianaNites** [comment]
What use-case does the `Implementors` section on `Traits` solve? Right now all it does is bloat pages and repeat the methods already listed above. Note that it doesn't repeat their documentation. Just a straight list of documentation-less method signatures that are already listed above, for everything that uses `Iterator`. 

The `Trait Implementations` section, which appears on structs, is fine as-is, and already does something different than the `Implementors` section, it has documentation.

**Comment by ebarnard** [comment]
When opening the Iterator page in the latest nightly docs (which include #56005) it takes about 45 seconds for "Loading content..." to disappear and the Iterator docs to become usable. This does not include network time. (Late 2013 MPB running Safari 12 and Firefox 64).

This renders the Iterator docs page unusable unless I am careful not to use the search feature, click on any links, or navigate away from the page.

I'm not sure whether others are experiencing similarly long load times but if they are it seems a convincing reason to revert #51885 until the performance issues can be fixed.

I'm also unsure of the intentions of the `Implementors` section. Is it meant to show docs written on methods in an `impl` block rather than on the methods in the `trait` declaration.

**Comment by GuillaumeGomez** [comment]
It takes 3 seconds for me, whatever the computer/OS.

**Comment by ebarnard** [comment]
Disabling extensions in Firefox brought it down to about 15s. Safari on Mac and iOS (iPhone 6S) is still about 45s.
Here's a Gecko profile for Firefox in case you're interested https://perfht.ml/2QBVmXi.

**Comment by GuillaumeGomez** [comment]
I'm on mac too (but never used safari though), strange. However, the profile tells that most of time is spent into rendering. So reducing DOM and discuss about removing trait implementations methods is still the good thing to do (and is being done already since a while).

**Comment by Stargateur** [comment]
10 minutes for my chrome. Also, try to avoid "work for me" argument, you seem to believe there is no problem here. I'm telling you that this will made crazy a lot of people who will use rust.

**Comment by GuillaumeGomez** [comment]
If I thought there were no problem, I wouldn't be working on it since I consider that a few seconds is still way too much. I was just saying that since you were using the same OS, having such rendering speed difference was strange.

**Comment by JeanMertz** [comment]
Thank you for the work you're doing on this, @GuillaumeGomez.

In case it helps, here's a recording of the request on my Mac that took about 40 seconds (without cache) until the screen was "painted" with some elements, and another 10 until text was shown.

https://www.icloud.com/iclouddrive/01XaTrKAGVgED_IAGNxerP6ig#doc.rust-lang.org.har

Also some attached screenshots of the inspector.

<img width="1112" alt="screenshot 2019-01-12 at 16 58 04" src="https://user-images.githubusercontent.com/383250/51075482-5f343400-168c-11e9-9622-344ab644b22b.png">

<img width="1112" alt="screenshot 2019-01-12 at 16 59 52" src="https://user-images.githubusercontent.com/383250/51075489-66f3d880-168c-11e9-81ec-aa482525c037.png">

<img width="1112" alt="screenshot 2019-01-12 at 16 58 33" src="https://user-images.githubusercontent.com/383250/51075494-6eb37d00-168c-11e9-8c38-cb1436b36373.png">

**Comment by GuillaumeGomez** [comment]
Wo, this is insane! A lot of work remains to be done in here. Thanks for the screens!

**Comment by leonlag** [comment]
I can confirm this on fedora 29. Using Chromium (without addons and new profile for testing) it's close to unusable(~3FPS) just scrolling through the page. Firefox is acceptable (~50FPS).

**Comment by carols10cents** [comment]
> A lot of work remains to be done in here.

I'm going to reopen this then, as a tracking issue of sorts? I think this could benefit from a checklist in the description of known issues and their status, perhaps with issues that need investigation as well?

Just a suggestion, because [I'm pointing some folks to this issue](), and it sounds like it isn't actually solved. Feel free to reclose and manage this in a different way!

**Comment by HarrisonMc555** [comment]
It is now March 1, 2019, and I can confirm that this is still an issue. Has there been more work on reducing the size of this page?

**Comment by GuillaumeGomez** [comment]
There was but no major improvement for the moment.

**Comment by kornelski** [comment]
The sheer size of it seems to be the problem. There's too much to load:

![webpagetest](https://user-images.githubusercontent.com/72159/55676885-85592100-58d5-11e9-87b4-c407ebe2fa71.png)

and too much for CSS to lay out:

<img width="780" alt="devtools" src="https://user-images.githubusercontent.com/72159/55676898-b5082900-58d5-11e9-9eb8-c81790f5961e.png">

I've never seen layout take so long, and I'm testing it on Intel's i9 CPU.

**Comment by memoryruins** [comment]
Layout takes around 15-17 seconds on my laptop for both stable/nightly versions of the docs.

![iter-docs-0](https://user-images.githubusercontent.com/6868531/55677122-20d79b00-58b0-11e9-8bad-8cc2e698ee4e.png)

CPU: Intel i5-4300U (4) @ 2.900GHz
Browser: Firefox 66.0.2 (64-bit)

Safari on an iPhone 5s takes roughly 50 seconds before loading text at the top of the page.

**Comment by samcday** [comment]
Just weighing in on my 2c worth here.

Just because it's 2019 doesn't mean it's reasonable to be shipping 12mb HTML files to users. Browsers today are super smart and fast and all, but actually they're not really much different to how they were 10+ years ago in one important sense: virtually all painting / reflow events can only take place on the main UI thread.

As a result, forcing the poor browser to parse 12mb is somewhat absurd. Jump on the std::iter::Iterator page and run this in the Javascript console:

```
```

**Comment by samcday** [comment]
I opened #59768 as a stopgap to alleviate this issue a bit.

It doesn't do anything to tackle the size of the Iterator doc page, but it does halve the time it takes to reach DOMContentLoaded on my machine. It does this by simply starting with all Implementors collapsed.

**Comment by kornelski** [comment]
Did something happen to rustdoc in general?

Source code for actix-web is 1.3MB, but the generated documentation takes 347MB!

https://github.com/actix/actix-web/issues/770

**Comment by GuillaumeGomez** [comment]
Not as far as I know. Normally it should be reducing size, not the opposite... Can you provide the list of files with their associated size please?

**Comment by ollie27** [comment]
@kornelski checking out https://github.com/actix/actix-web/commit/3e986042973762e28fa6af25c8d161ce9f4de4d0 locally I'm only measuring 249MiB and I don't see anything too unusual. It includes documentation for all dependencies including things like `libc` which is 20MiB alone. The trait pages are very big like `futures/stream/trait.Stream.html` being 2.2MiB which is caused by the regressing caused by #51885 that this issue is tracking.

Can someone add a `regression-from-stable-to-stable` label to this issue?

**Comment by lzutao** [comment]
As you wish @ollie27 
@rustbot modify labels: +regression-from-stable-to-stable

**Comment by ollie27** [comment]
Maybe that will work for me.

@rustbot modify labels: +regression-from-stable-to-stable

**Comment by insanitybit** [comment]
https://rusoto.github.io/rusoto/rusoto_dynamodb/struct.PutItemInput.html

I've experienced serious lag/ hangs/ CPU spikes on rusoto docs.

**Comment by wafflespeanut** [comment]
I have a suggestion related to trait pages (open to bikeshedding).

Most often, we're only interested in *whether* a type implements a trait (when we're in some trait's page) or finding a trait implementation in a type's page. Does it make sense to get rid of the detail on trait implementations (under "Implementations on foreign types" and "Implementors" sections)?

Basically, these:

![trait-impl](https://user-images.githubusercontent.com/6691262/56808686-af0bc500-684f-11e9-9d77-94380233dcc5.png)

For example, let's take the impl for `Args` [shown in `Iterator` trait](https://doc.rust-lang.org/nightly/std/iter/trait.Iterator.html#implementors) page. It's very useful to link to the source of that implementation, but it's probably not that useful to show the implementation detail with all those methods because we're already on the trait's page?

![example](https://user-images.githubusercontent.com/6691262/56809329-360d6d00-6851-11e9-9fdc-4d26d98c6978.png)

---

Instead, we could link to the trait impl in the type's page. Maybe an icon next to that impl? (again, this is just a suggestion, we could go for something else). The user can find the relevant impl information in the type's page.

![alt](https://user-images.githubusercontent.com/6691262/56810026-ed56b380-6852-11e9-85cb-71c2b00aa70d.png)

This should improve the performance for trait pages (particularly Iterators) because we'll be cutting a huge chunk of nodes in the DOM.

**Comment by kornelski** [comment]
BTW, all content under these `[+]`'s is duplicate. You get all the same iterator  methods repeated over and over again for every `Itearator` impl. It'd make sense to group implementations by trait, and include methods *once*.

**Comment by Manishearth** [comment]
I think that somewhat makes sense for common traits like Iterator, but not in general.

One useful side of the expanded doc view is being able to see what methods you have available. Often these are through traits.

**Comment by Stargateur** [comment]
Note that @shepmaster give me a trick, use a past version of the doc, https://doc.rust-lang.org/1.29.0/std/iter/trait.Iterator.html. Not perfect but that do the work until the problem is solved.

**Comment by matthieu-m** [comment]
> 
> 
> I think that somewhat makes sense for common traits like Iterator, but not in general.
> 
> One useful side of the expanded doc view is being able to see what methods you have available. Often these are through traits.

There are two distinct use cases though:

 1. Showing which traits are implemented by the current struct/trait A on the page of A, and the methods available.
 2. Showing all implementations of a trait A on the page of A, and duplicate the trait methods for each implementation.

I agree with you that (1) is very useful, however (2) is a completely separate usecase and the cause of the issue for `Iterator`.

I find that @wafflespeanut adequately solves both (2) and the bloat issue: it shows all implementers, yet conveniently sidesteps the performance issue by referring to the implementer's page for further information.

**Comment by Manishearth** [comment]
Oh, I see. Hmm. That seems okay.

(Though I'll note the performance issues are due to various bugs and not the actual page content. Browsers are pretty good about not processing display:none things)

**Comment by wafflespeanut** [comment]
> One useful side of the expanded doc view is being able to see what methods you have available. Often these are through traits.

This is true and this is the case for "type" pages. I never suggested to remove anything there. I was talking about removing the stuff in "trait" pages, because we already have all the trait methods up top (so why duplicate everything?). I'm also suggesting to provide a link (next to the trait impl in a trait's page) for the user to see that impl in the corresponding type's page.

> Though I'll note the performance issues are due to various bugs and not the actual page content. 

I agree that browsers are very good, but I still think the current state of rustdoc (specifically for Iterator trait) is an overkill, especially when we don't have a good reason for listing all trait methods for all trait impls.

**Comment by sinesc** [comment]
As soon as the last "Loading content..." appears the browser has processed the entire HTML. After that it is "just" layouting the blocks that were previously display: none.
One could flip those blocks individually from hidden to visible with a setTimeout(.. ,0) inbetween to allow the browser to layout the blocks individually. Total rendering time would go up slightly, but the stuff at the top would appear faster.

**Comment by Manishearth** [comment]
> I agree that browsers are very good, but I still think the current state of rustdoc (specifically for Iterator trait) is an overkill, especially when we don't have a good reason for listing all trait methods for all trait impls.

Yeah I agree with that.

I'm more saying that this isn't the cause of the performance issues.

**Comment by GuillaumeGomez** [comment]
It could be part of it. That's something that has been suggested a while ago and that I wanted to give a try to. I'll try to find some time to check locally if that improves things.

**Comment by bendem** [comment]
Speaking of doc performance, you might want to fix the server's config to compress html documents. Those 13MB are 692K gzip'd, 74K brotli'd.

**Comment by GuillaumeGomez** [comment]
That's not up to rustdoc unfortunately. Please open an issue on docs.rs repository.

**Comment by ebarnard** [comment]
The `Iterator` docs are hosted on https://doc.rust-lang.org. Presumably that isn't hosted by docs.rs?

**Comment by GuillaumeGomez** [comment]
Like I said, on their repository: https://github.com/rust-lang/docs.rs

**Comment by SimonSapin** [comment]
I believe the hosting of `rust-lang.org` (including `doc.rust-lang.org`) is the responsibility of the @rust-lang/infra team. Maybe a new issue on https://github.com/rust-lang/infra-team/issues would be the appropriate place to discuss server configuration?

**Comment by GuillaumeGomez** [comment]
Arf sorry, I kept reading `docs.rs` instead of `doc.rust-lang.org`...

**Comment by pietroalbini** [comment]
> Speaking of doc performance, you might want to fix the server's config to compress html documents. Those 13MB are 692K gzip'd, 74K brotli'd.

We already have automatic compression configured on the CloudFront side, but CloudFront won't compress files bigger than 10MB, so fixing this is not a simple configuration change.

**Comment by ebarnard** [comment]
Can I suggest a minimal change, as far as I can see without downsides, that should significantly reduce the size of the Iterator doc page:

In the Implementors and Implementations on Foreign Types sections, only show methods that appear in the `impl` block for that type. This has the benefit of
- Reducing the size of the Iterator page, and other large trait documentation pages.
- Retaining documentation on the `impl` blocks and functions in the `impl` blocks.
- Additionally indicating which provided methods are overridden.
- Being a small change that can be easily backed out if issues arise.

What this would look like for the `Read` trait:
<img width="1428" alt="Screenshot 2019-06-03 at 16 41 27" src="https://user-images.githubusercontent.com/1059683/58814907-84ffac80-861e-11e9-8692-79be473a5299.png">

This should be an easy change to make so I'll try and have a PR ready soon so we can see what impact it would have on page sizes and load times.

**Comment by ebarnard** [comment]
A set of Rust stdlib docs build with this change are [available here](https://ebarnard.github.io/2019-06-03-rust-smaller-trait-implementers-docs/).

[The `Iterator` doc page](https://ebarnard.github.io/2019-06-03-rust-smaller-trait-implementers-docs/std/iter/trait.Iterator.html) is now 724kB.

**Comment by memoryruins** [comment]
Thank you for trying these changes and providing a demo!

Immediate impression:

- Current nightly Iterator docs: similar to my previous report <https://github.com/rust-lang/rust/issues/55900#issuecomment-480550869>,
  except causes my mobile browser to become unresponsive and crash.
- @ebarnard 's iterator [docs](https://ebarnard.github.io/2019-06-03-rust-smaller-trait-implementers-docs/std/iter/trait.Iterator.html): loads nearly immediately on laptop 🎉, and it scrolls smoothly without page flickering or laptop fans becoming loud. It loaded quickly on my phone too :)
  It appears to make docs pages clearer and easier to navigate in general.

Another example trait page that should benefit, which I'm unable to load on my laptop:
https://docs.rs/uom/0.23.1/uom/trait.Conversion.html

**Comment by GuillaumeGomez** [comment]
@ebarnard: Looks great! Can you open a PR with your changes please? It'll also allow to see if other people like this change or not.

**Comment by ebarnard** [comment]
@GuillaumeGomez It's open at #61505

**Comment by GuillaumeGomez** [comment]
Little update: I think I'll investigate wasm at some point and see if it improves the situation. But first, I want to check if there is a way to improve how we store the information in the `search-index.js` file.

**Comment by ollie27** [comment]
I think this issue should be closed. The original regression caused by #51885 has long since been fixed by #61505. I think we can track specific performance problems in dedicated issues like #56545.
